python/dask_cudf/dask_cudf/io/parquet.py

```diff
@@ -1,58 +1,247 @@
 # Copyright (c) 2024, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
 import functools
+import itertools
+import math
+import os
+import warnings
+from typing import TYPE_CHECKING, Any
 
+import numpy as np
 import pandas as pd
-from dask_expr.io.io import FusedParquetIO
-from dask_expr.io.parquet import FragmentWrapper, ReadParquetPyarrowFS
+from dask_expr._expr import Elemwise
+from dask_expr._util import _convert_to_list
+from dask_expr.io.io import FusedIO, FusedParquetIO
+from dask_expr.io.parquet import (
+    FragmentWrapper,
+    ReadParquetFSSpec,
+    ReadParquetPyarrowFS,
+)
 
 from dask._task_spec import Task
+from dask.dataframe.io.parquet.arrow import _filters_to_expression
+from dask.dataframe.io.parquet.core import ParquetFunctionWrapper
+from dask.tokenize import tokenize
+from dask.utils import parse_bytes
 
 import cudf
 
-from dask_cudf import _deprecated_api
+from dask_cudf import QUERY_PLANNING_ON, _deprecated_api
 
 # Dask-expr imports CudfEngine from this module
 from dask_cudf._legacy.io.parquet import CudfEngine  # noqa: F401
 
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+
+_DEVICE_SIZE_CACHE: int | None = None
+
+
+def _get_device_size():
+    try:
+        # Use PyNVML to find the worker device size.
+        import pynvml
+
+        pynvml.nvmlInit()
+        dev = int(os.environ.get("CUDA_VISIBLE_DEVICES", "0").split(",")[0])
+        handle = pynvml.nvmlDeviceGetHandleByIndex(dev)
+        return pynvml.nvmlDeviceGetMemoryInfo(handle).total
```

Comment:
@pentschev - Just want to double check with you that this is a reasonable way to measure the size of a single device for each worker?

Reply:
Yes, this is correct at least assuming you only need the device's full memory capacity (i.e., not free memory) and support for querying devices by index only is sufficient. The above won't work for MIG or querying by UUID, for example. If you need more robust support you'll need something more elaborate like [this](https://github.com/rapidsai/dask-cuda/blob/9e7a926bc305f32aa0463a44eb96d4494fe55fc0/dask_cuda/utils.py#L154-L163).

Reply:
We also have that exact function you want [here](https://github.com/rapidsai/dask-cuda/blob/9e7a926bc305f32aa0463a44eb96d4494fe55fc0/dask_cuda/utils.py#L185-L197), if you're ok with a Dask-CUDA dependency that may suffice.

Reply:
Thanks @pentschev ! Just to clarify, I want to run a function on every worker that returns the total size of the device it will use for executing Dask-cuDF tasks. Would running `get_device_total_memory` on all workers return this information (i.e. all workers target `index=0`)?

Reply:
Aha, I think I see. I still need to check/split `CUDA_VISIBLE_DEVICES` to get the index, but I should use the MIG-compatible logic.

Reply:
Yes, you can do that. As you noticed you'll need to split `CUDA_VISIBLE_DEVICES` and use only the first element (the active GPU). This presumes, of course, you're running Dask-CUDA where `CUDA_VISIBLE_DEVICES` is always properly set and won't work with vanilla Distributed out-of-the-box. And be aware only the total GPU memory size is reliable, checking free/used memory isn't due to the common use of RMM pool.